### PR TITLE
fix(spec-decode): separate AR draft Philox stream

### LINKED ISSUE #205: [correctness] MTP probabilistic draft Gumbel noise overlaps verifier Philox range (V2 runner)
Provenance: found during a source review of the v20 r12 release image `voipmonitor/vllm:gilded-gnosis-v20-vllmb46c3aa-si35aebc6-fi801d57a-cu132-20260730-r12`; code refs are against the composed r12 tree (`r12-composed`). Permalinks point at the last commit touching each file (blobs are byte-identical to the composed tree): speculators @ `467582afc`, `spec_decode/utils.py` @ `d179dc837`, `rejection_sampler_utils.py` / `sample/gumbel.py` @ `b453c659b`.

Deployment context: GLM-5.2 753B MoE, MTP draft, TP4, V≈154k, per-request seeds. Affected configuration: **V2 runner (`VLLM_USE_V2_MODEL_RUNNER=1`) + `draft_sample_method="probabilistic"` + sampled requests (temperature > 0)** — i.e. the release-gate MTP-probabilistic path. Greedy drafts and the V1 runner are unaffected (different code). This is a correctness/accuracy concern, not memory; proposal from a static review, happy to be corrected.

## Summary

`AutoRegressiveSpeculator.sample_draft` (inherited by `MTPSpeculator`, `EagleSpeculator`, `Gemma4Speculator`) keys the draft-proposal Gumbel noise with `positions + 1`, while the base class and every other speculator use the salted `draft_gumbel_pos(positions)` (= `positions + 1 + 2**30`) precisely so the draft stream stays disjoint from the rejection sampler's Philox range. With the unsalted override, the draft proposal noise and the verifier's acceptance/recovery/bonus noise are drawn from the **same (seed, offset) Philox pairs** every decode iteration, which breaks the independence assumptions of Leviathan-style rejection sampling.

## Where

**The override** — [`vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L370-L387`](https://github.com/local-inference-lab/vllm/blob/467582afcaff9ff12b44fd03240cf8b400eba489/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L370-L387):

```python
logits = self.model.compute_logits(hidden_states)
if draft_logits is not None:
    # NOTE(woosuk): We must add 1 to the positions to match the Gumbel noise
    # used for draft and target sampling.
    return gumbel_sample(
        logits,
        idx_mapping,
        temperature,
        seeds,
        positions + 1,
        ...
```

**The documented contract it bypasses** — [`vllm/v1/worker/gpu/spec_decode/utils.py#L10-L23`](https://github.com/local-inference-lab/vllm/blob/d179dc83755ca7365a6c1b1294c74d7908106bc7/vllm/v1/worker/gpu/spec_decode/utils.py#L10-L23):

```python
# Salt added to the Philox offsets used for draft-token Gumbel noise.
# Positions are bounded by max_model_len, so this puts the draft stream in a
# range disjoint from the target-side offsets.
DRAFT_GUMBEL_POS_OFFSET = 1 << 30


def draft_gumbel_pos(positions: torch.Tensor) -> torch.Tensor:
    """Philox offsets for draft Gumbel noise, given draft-row positions.

    The rejection sampler keys both its acceptance uniform and recovery
    Gumbel noise for position P by offset P. Keep the draft proposal on a
    disjoint Philox range so rejection sampling remains unbiased.
    """
    return positions + (1 + DRAFT_GUMBEL_POS_OFFSET)
```

The base class uses it ([`speculator.py#L366-L379`](https://github.com/local-inference-lab/vllm/blob/467582afcaff9ff12b44fd03240cf8b400eba489/vllm/v1/worker/gpu/spec_decode/speculator.py#L366-L379): `gumbel_sample(..., draft_gumbel_pos(positions), ...)`), and DFlash/DSpark go out of their way to honor it (`dflash/speculator.py:461-463`, `dspark/speculator.py:255-265`: "…so pass Q-2 to get a key unique per predicted position **and disjoint from the rejection sampler's acceptance/recovery keys**"). Only the autoregressive override drops the salt.

## Mechanism — the streams do collide

**Same seed tensor.** Both sides use the identical per-request seeds: the runner passes `self.sampler.sampling_states.seeds.gpu` into `speculator.propose` (`model_runner.py:2032-2033`), the speculator copies it (`speculator.py:397-398`), and the rejection sampler receives the same tensor (`gpu/spec_decode/rejection_sampler.py:148-149`). Both index it by the same `req_state_idx`.

**Same offset range.** The V2 rejection sampler keys its randomness by the **absolute row position** (`pos = input_batch.positions[logits_indices]`, `gpu/spec_decode/rejection_sampler.py:130` → `rejection_sample(..., pos, ...)` at `:144`):

- acceptance uniform for the i-th draft token, [`rejection_sampler_utils.py#L530-L534`](https://github.com/local-inference-lab/vllm/blob/b453c659b7218b77690cbc3e343814c9b1d98780/vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py#L530-L534):

```python
pos = tl.load(pos_ptr + logit_idx)
u = tl_rand32(seed, pos, includes_zero=False)
```

- recovery/bonus Gumbel, `_resample_kernel` → `gumbel_block_argmax` with the same `pos_ptr` (`rejection_sampler_utils.py:776-793`), which derives ([`gumbel.py#L147-L158`](https://github.com/local-inference-lab/vllm/blob/b453c659b7218b77690cbc3e343814c9b1d98780/vllm/v1/worker/gpu/sample/gumbel.py#L147-L158)):

```python
seed = tl.load(seeds_ptr + req_state_idx, ...)
pos = tl.load(pos_ptr + token_idx)
gumbel_seed = tl.randint(seed, pos)
...
u = tl_rand32(gumbel_seed, block, includes_zero=False)
```

Let the verify batch for a request be rows `[t@R, d1@R+1, …, dk@R+k]` (R = position of the last committed token). The verifier consumes offsets `{R, …, R+k}`: acceptance `u` for `d_{i}` at offset `R+i-1`; recovery Gumbel after `a ≥ 1` acceptances at offset `R+a`; bonus Gumbel at offset `R+k`. The draft, meanwhile, proposed `d_j` (which sits at position `R+j`) with key `positions+1 = R+j` — inside the same `{R+1, …, R+k}` window, same seed. Concretely, **every decode iteration**:

1. **Acceptance uniform ← draft seed word.** Triton's `tl.rand(seed, off)` is `uint32_to_uniform_float(tl.randint(seed, off))` — the same Philox word. So the acceptance uniform for `d_{j+1}` (offset `R+j`) is the float reinterpretation of the exact 32-bit word used as `gumbel_seed` for proposing `d_j`. The uniform that decides whether to accept a token is a deterministic function of the noise that generated the previous, already-in-context draft token — the u ⟂ (draft, distributions) independence in the Leviathan et al. proof no longer holds.
2. **Recovery noise ≡ draft proposal noise.** On rejection after `a ≥ 1` acceptances, the recovery Gumbel field is `tl_rand32(tl.randint(seed, R+a), v)` over vocab `v` — bit-identical to the field the draft used to propose `d_a` (same construction, same global vocab indexing on both sides). The residual `max(p−q, 0)` resample is performed with the same noise realization that selected the last accepted token.
3. **Bonus noise ≡ last draft's proposal noise.** When all k drafts are accepted, the bonus token at offset `R+k` reuses `d_k`'s exact Gumbel field: bonus = `argmax(p/T + g)` where `d_k = argmax(q/T + g)` with the *same* g. When the target and draft distributions are close (they are, for a good MTP head — that's the point), this correlates the bonus token with the preceding draft token far beyond what independent sampling would produce (a plausible symptom is repetition amplification in sampled output).

Note the override's comment says the +1 is "to match the Gumbel noise used for draft and target sampling", but no valid coupling is achieved either: the target-side convention keys the token at position Q by offset Q−1 (main sampler `pos = positions[logits_indices]`; verifier as above), while the override keys it by Q. The result is not a coupled-sampling scheme — it's an off-by-one collision with the verifier's keys for neighboring positions.

## Suggested fix

One line — delegate to the documented contract in the override:

```python
return gumbel_sample(
    logits, idx_mapping, temperature, seeds,
    draft_gumbel_pos(positions),   # was: positions + 1
    ...
)
```

(or simply drop the override's probabilistic branch and call `super().sample_draft(...)`, keeping the greedy short-circuit). `draft_gumbel_pos` keeps keys unique per predicted position and disjoint from the verifier by construction. History suggests this is an incomplete migration: the salt was introduced in `d179dc837` in the base class, and the autoregressive override kept the pre-salt `positions + 1`.

## Verification hint

- Instrument both sides and log `(seed, offset)` pairs for one sampled request with MTP probabilistic on the V2 runner: the draft's `gumbel_sample` offsets equal the verifier's acceptance/recovery offsets shifted by one position, in the same iteration.
- Statistical A/B: fixed seed, temperature 1, compare the empirical distribution of bonus tokens (and rejected-position recovery tokens) against pure target sampling, before/after the one-line fix; the unsalted variant shows excess `P(bonus == d_k)`.

Cross-ref: memory-side findings for the same path are in #200.

### PR DESCRIPTION
**Pull request by malaiwah**
## Summary

Move probabilistic proposal noise in the V2 autoregressive speculator onto the documented salted draft Philox range, so rejection recovery cannot reuse the exact Gumbel field that selected the rejected draft token.

Fixes #205.

## Corrected scope

The issue's re-review established the exact live surface:

- affected: V2 `AutoRegressiveSpeculator` with probabilistic drafts and sampled requests;
- unaffected: greedy drafts, the V1 proposer, DFlash/DSpark, and the all-accepted bonus draw;
- shipped live preset: plain `glm52` / `glm52-hybrid` with `MTP>0`, V2, default probabilistic draft sampling;
- r12 EXL3 release gate is greedy and the production turnkey uses V1, so neither executes this line.

The first-order defect is recovery after rejection. The verifier reuses the rejected draft token's unsalted `(seed, offset)` and therefore the exact per-vocabulary Gumbel vector conditioned on that draft having won. Acceptance-word correlation is negligible; the bonus offset is already disjoint.

## Change

Replace:

```python
positions + 1
```

with:

```python
draft_gumbel_pos(positions)
```

in `AutoRegressiveSpeculator.sample_draft`.

The override remains intact. In particular, it still forwards `output_processed_logits_active_rows=self.active_num_reqs`, which prevents padded FULL-graph rows from overwriting real request state.

## Invariants

- Draft proposal offsets become `positions + 1 + 2**30`, exactly disjoint from verifier offsets.
- The existing int64 tensor shape and one elementwise add are unchanged; only the immediate changes.
- FULL-graph active-row protection is unchanged.
- Greedy behavior, bonus offsets, acceptance logic, capture operation count, allocations, public ABI, and non-AR speculators are unchanged.
- Output tokens can change bit-for-bit on the affected probabilistic path; distributionally this restores the rejection sampler's independence contract.

## Verification

CPU-isolated r12 container:

```text
3 passed
```

The new regression constructs the real AR override with probabilistic draft logits, intercepts `gumbel_sample`, and asserts the exact passed position tensor is:

```text
positions + 1 + 2**30
```

Additional checks:

- Ruff lint and format: pass
- `git diff --check`: pass
- independent adversarial review: APPROVE, confidence 0.99, no findings

The reviewer re-derived the draft/rejection offset algebra and confirmed the rejected-token collision is removed while bonus range, FULL-graph safeguards, dtype/shape, and greedy behavior remain unchanged.

No acceptance-rate or user-visible distribution magnitude claim is made without the optional statistical A/B. Production was not rebuilt or restarted.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved speculative decoding’s probabilistic draft sampling to use separate random-number generator offsets, preventing overlap with acceptance and recovery sampling.
  * Deterministic draft sampling behavior remains unchanged.

* **Tests**
  * Added coverage verifying that probabilistic draft sampling uses disjoint random-number generator offsets.

### CODE REVIEW

**Comment by voipmonitor** [comment]
Superseded by #211. The replacement keeps the same salted Philox correctness fix, centralizes probabilistic draft sampling in the base speculator, routes the GG autoregressive and DSpark paths through it, and preserves the autoregressive FULL-graph active-row safeguard. Merge #211 only.
